### PR TITLE
Add missing clang-tidy fixes for modernize-use-equals-(default|delete)

### PR DESCRIPTION
**Pull request by Skylion007**
More clang-tidy for default or deleting more ctors and dtors. This is slightly more efficient and more readable

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

### CODE REVIEW

**Review by ezyang — APPROVED** [review]

**Comment by lezcano** [comment]
Rather than fixing these as one-offs, could it be possible to add the relevant rule to `lintrunner`?

**Comment by Skylion007** [comment]
> Rather than fixing these as one-offs, could it be possible to add the relevant rule to `lintrunner`?

It's unfortunately more complicate than that as it's difficult to configure (more general lint rules) with clang-tidy 11. This is further complicated by issues with the header-filter regex (POSIX only support, no support for look-aheads). There are several open issues on trying to expand clang-tidy coverage, but it's a bit tricky at the moment. This is further complicated by ensuring compatibility with the internals only version of clang-tidy.

**Comment by Skylion007** [comment]
> Rather than fixing these as one-offs, could it be possible to add the relevant rule to `lintrunner`?

TLDR: the rule is already in the linter, but the clang-tidy linter is only applied to a fraction of the files.

**Comment by ezyang** [comment]
@malfet will upload the latest clang-tidy binary to S3 so you can start on the migration process to latest clang-tidy.

File: torch/csrc/distributed/autograd/context/container.h [hunk 1]
```diff
@@ -117,7 +117,7 @@ class TORCH_API DistAutogradContainer {
     std::unordered_map<int64_t, ContextPtr> contexts;
   };
 
-  DistAutogradContainer();
+  DistAutogradContainer() = delete;
```
Comment (malfet) [inline torch/csrc/distributed/autograd/context/container.h:120]: Wow, this is neat. So before that it was up to linker to check that DistAutogradContainer is not constructed by default? 8)

**Comment by malfet** [comment]
> @malfet will upload the latest clang-tidy binary to S3 so you can start on the migration process to latest clang-tidy.

I'm not going to do that, but rather update the workflows to build it, see https://github.com/pytorch/test-infra/pull/1382
